LANGUAGE_SDK_SPECIFICATION.md

```diff
@@ -0,0 +1,1325 @@
+
+#### 4.6.1 Purpose
+
+Enable external systems to complete an operation via API calls.
+
+#### 4.6.2 Required Behavior
+
+The SDK MUST:
+
+1. Create a CALLBACK operation and retrieve its `CallbackId`
+2. Provide the callback ID to the user for external completion
+3. Support optional timeout and heartbeat timeout configuration
+4. Handle completion via `SendDurableExecutionCallbackSuccess` or `SendDurableExecutionCallbackFailure`
+5. Support heartbeat via `SendDurableExecutionCallbackHeartbeat`
+
+#### 4.6.3 Supported Actions
+
+- `START`: Create a callback waiting for external completion
+
+#### 4.6.4 State Transitions
+
+```
+[New] → START → STARTED → (external success) → SUCCEEDED [Done]
+                        ↓
+                        └→ (external failure) → FAILED [Done]
+                        └→ (timeout) → TIMED_OUT [Done]
+```
+
+#### 4.6.5 Configuration Options
+
+The SDK SHOULD support:
+
+- `TimeoutSeconds`: Maximum time to wait for callback completion
+- `HeartbeatTimeoutSeconds`: Maximum time between heartbeat signals
+
+### 4.7 CHAINED_INVOKE Operation
+
+#### 4.7.1 Purpose
+
+Invoke another Lambda function (durable or non-durable) and wait for its completion.
+
+#### 4.7.2 Required Behavior
+
+The SDK MUST:
+
+1. Accept function name or ARN
+2. Serialize input payload
+3. Create a CHAINED_INVOKE operation with the target and payload
+4. Handle both durable and non-durable target functions
+5. Return the target function's result upon success
+6. Propagate errors from the target function
+
+#### 4.7.3 Supported Actions
+
+- `START`: Invoke the target function
+
+#### 4.7.4 State Transitions
+
+```
+[New] → START → STARTED → (invoke completes) → SUCCEEDED [Done]
+                        ↓
+                        └→ (invoke fails) → FAILED [Done]
+                        └→ (invoke timeout) → TIMED_OUT [Done]
+                        └→ (invoke stopped) → STOPPED [Done]
+```
+
+#### 4.7.5 Configuration Options
+
+The SDK MUST support:
+
+- `FunctionName`: Name or ARN of the target Lambda function
+
+The SDK MAY support:
+
+- `TenantId`: For tenant isolation scenarios
+
+### 4.8 CONTEXT Operation
+
+#### 4.8.1 Purpose
+
+Create parent-child operation relationships for isolation, concurrency control, and replay optimization.
+
+#### 4.8.2 Required Behavior
+
+The SDK MUST:
+
+1. Support nesting operations within a context
+2. Associate child operations with the context via `ParentId`
+3. Propagate context completion to all children
+4. Support conditional replay of children
+5. Enable context-level error handling
+
+#### 4.8.3 Supported Actions
+
+- `START`: Begin a new context
+- `SUCCEED`: Complete the context successfully
+- `FAIL`: Complete the context with an error
+
+#### 4.8.4 State Transitions
+
+```
+[New] → START → STARTED → SUCCEED → SUCCEEDED [Done]
+                        ↓
+                        └→ FAIL → FAILED [Done]
+```
+
+#### 4.8.5 Replay Optimization
+
+The SDK SHOULD support `ReplayChildren` option:
```

Comment:
The SDK doesn't need to implement any optimization but it does need to pass this in the API if it _wants_ to have its children's state included when replaying. Otherwise the default is *not* to include the children.

The JS SDK uses this to support `parallel`/`map` that have a total size larger than the individual operation size limit (it will ask for the children state and replay each branch to combine the output)